GH Issue #15652 SqlFileStream Implementation

keeratsingh opened the pull request:
This Pull Request ports the SqlFileStream #15652 capability only on Windows.
SqlClient throws `SqlFileStream is not supported on this platform` exception for Non-Windows Platforms.

| Client        | Server           | Feasible  | Comments  |
| :------------- |:-------------| :-----|  :----- |
| Windows      | Windows | Yes | Implemented by this PR. |
| Windows      | Linux      |  No |  SQL Server on Linux does not support FileStream. |
| Linux | Windows      |   No | Currently there is no way to open a file over SMB and pass the Extended Attributes. |
| Linux | Linux      |    No | SQL Server on Linux does not support FileStream. |

keeratsingh commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -4,13 +4,15 @@
     {
         // https://msdn.microsoft.com/en-us/library/bb432380.aspx
         // https://msdn.microsoft.com/en-us/library/windows/hardware/ff566424.aspx
+#pragma warning disable BCL0015 // Invalid Pinvoke call
```
Suppressing warnings as per Jan's suggestion #29169

JeremyKuhne replied:
Please remove the pragma- shouldn't need it if we aren't including this on UAP.

---
In reply to: [190421828](https://github.com/dotnet/corefx/pull/29879#discussion_r190421828) [](ancestors = 190421828)

keeratsingh commented on src/Common/src/Interop/Windows/NtDll/Interop.RtlNtStatusToDosError.cs:
```diff
@@ -10,8 +10,10 @@ internal partial class Interop
     internal partial class NtDll
     {
         // https://msdn.microsoft.com/en-us/library/windows/desktop/ms680600(v=vs.85).aspx
+		#pragma warning disable BCL0015 // Invalid Pinvoke call
```
Suppressing warnings as per Jan's suggestion #29169

JeremyKuhne replied:
You should not need this if you're only including the files in the non UAP builds. For example:



https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System.IO.FileSystem.csproj#L9

https://github.com/dotnet/corefx/blob/master/src/System.IO.FileSystem/src/System.IO.FileSystem.csproj#L187-L216

keeratsingh replied:
@JeremyKuhne 
Adding the condition `'$(UWPCompatible)' != 'true'` as per the suggestion throws an error `TypesMustExist : Type 'System.Data.SqlTypes.SqlFileStream' does not exist in the implementation but it does exist in the contract.`

keeratsingh commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -33,7 +36,23 @@ internal partial class NtDll
+        {
+                return ( CreateFile(eaBuffer: new Microsoft.Win32.SafeHandles.SafeFileHandle(IntPtr.Zero, true), eaLength: 0, path, rootDirectory,createDisposition, desiredAccess, shareAccess, fileAttributes, createOptions, objectAttributes) );
+        }
+        internal unsafe static (int status, IntPtr handle) CreateFile(
```
Had to create a separate CreateFile Method as SqlFileStream requires the Extended Attributes buffer to be supplied and currently there is currently no way to do so other than NTCreateFile #29379

saurabh500 replied:
Nit: New line before the method.

stephentoub commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            return path;
+        }
+
+        static private void DemandAccessPermission
```
All of this CAS-related stuff can be deleted.

keeratsingh replied:
Is CAS-related stuff no longer valid/required ?

stephentoub replied:
Right, it's all nops on core.

stephentoub commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Unix.cs:
```diff
@@ -0,0 +1,69 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
This file looks identical to the .Uap.cs one.  You could have a single SqlFileStream.Unsupported.cs file that's included in both the UAP and Unix builds.

That said, why can't it be supported?  There's no way to make the majority of the functionality work on Unix or in UAP?

keeratsingh replied:
I will create a singe file `SqlFileStream.Unsupported.cs`
As to why can't it be supported answered in the comments below.

stephentoub commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+                try
+                {
+
+                    (int status, IntPtr handle) = Interop.NtDll.CreateFile( 
```
Why do we need to use Win32 interop for this?  i.e. why can't this just use FileStream?

keeratsingh replied:
As per reasons mentioned in #29379, since it is not a real UNC path, the bogus path only has meaning in the context of this transaction, hence we cannot use FileStream directly.

stephentoub commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+                bool bRevertAssert = false;
+                try
+                {
+                    SecurityPermission sp = new SecurityPermission(SecurityPermissionFlag.UnmanagedCode);
```
All of the CAS-related code can be deleted.

keeratsingh commented on src/System.Data.SqlClient/tests/ManualTests/SQL/SqlFileStreamTest/SqlFileStreamTest.cs:
```diff
@@ -0,0 +1,217 @@
+                        byte[] transactionContext = reader.GetSqlBytes(1).Buffer;
+                        
+                        // Create the SqlFileStream  
+                        using (Stream fileStream = new SqlFileStream(path, transactionContext, FileAccess.Read, FileOptions.SequentialScan, allocationSize: 0))
```
Explained in the comment above and more information in comments #15652
Also, this is Windows-only because SQL Server on Linux doesn't support `FileStream` currently.
In case of Linux Client and Windows SQL Server, as per discussions with Robert Dorr, currently there is no way to open a file over SMB and pass the Extended Attributes on Linux.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.FILE_FULL_EA_INFORMATION.cs:
```diff
@@ -0,0 +1,45 @@
+			/// <summary>
+            /// The offset of the next FILE_FULL_EA_INFORMATION-type entry. This member is zero if no other entries follow this one.
+            /// </summary>
+            internal UInt32 nextEntryOffset;
```
nit: we use the C# keywords for types (uint, byte, byte, ushort, byte)

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -101,7 +124,7 @@ public unsafe struct OBJECT_ATTRIBUTES
             public unsafe OBJECT_ATTRIBUTES(UNICODE_STRING* objectName, ObjectAttributes attributes, IntPtr rootDirectory)
             {
-                Length = (uint)sizeof(OBJECT_ATTRIBUTES);
+                Length = (uint)Marshal.SizeOf(typeof(OBJECT_ATTRIBUTES));
```
Don't change this to Marshal.SizeOf- it is much more expensive.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.SECURITY_QUALITY_OF_SERVICE.cs:
```diff
@@ -0,0 +1,42 @@
+			/// <summary>
+            /// Specifies the size, in bytes, of this structure.
+            /// </summary>
+            internal UInt32 length;
```
uint

JeremyKuhne replied:
Also, please match the casing of the docs. (`Length`, etc.)

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Access.cs:
```diff
@@ -0,0 +1,41 @@
+		
+        public enum Access
+        {
+			/// <summary>
```
You've got tabs in here, please use spaces for indenting

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Access.cs:
```diff
@@ -0,0 +1,41 @@
+    internal partial class Kernel32
+    {
+		/// <summary>
+        /// <a href="https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/buffer-descriptions-for-i-o-control-codes">RequiredAccess</a>.
```
Why this link?

keeratsingh replied:
Thank you for catching this, must have mixed up the links. I have updated the link with the second link, since the first link throws error '404 Content not found'

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Access.cs:
```diff
@@ -0,0 +1,41 @@
+		/// Indicates the type of access that a caller must request when opening the file object that represents the device (see IRP_MJ_CREATE).
+        /// </summary>
+		
+        public enum Access
```
You can make this a `[Flags]` to represent that you can OR the values. The other option (given that `Any` is `0`), is to define `FILE_READWRITE_ACCESS` with a comment noting that it isn't actually defined in Windows, but we're including it for convenience.

keeratsingh replied:
I have added the `[Flags]` attribute to the enum.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Access.cs:
```diff
@@ -0,0 +1,41 @@
+			/// <summary>
+            /// The I/O manager sends the IRP for any caller that has a handle to the file object that represents the target device object.
+			/// </summary>
+            FILE_ANY_ACCESS,
```
Should call out that this value should be avoided. (https://docs.microsoft.com/en-us/previous-versions/windows/hardware/design/dn614603) Also, consider actually putting in the defined values explicitly (i.e. 0x00, 0x01, etc.)

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Access.cs:
```diff
@@ -0,0 +1,41 @@
+		/// Indicates the type of access that a caller must request when opening the file object that represents the device (see IRP_MJ_CREATE).
+        /// </summary>
+		
+        public enum Access
```
ControlCodeAccess or IoControlCodeAccess would be a better name for the enum.

keeratsingh replied:
Agreed, updated the name to `IoControlCodeAccess` as suggested, also updated the filename to reflect the same.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.SECURITY_QUALITY_OF_SERVICE.cs:
```diff
@@ -0,0 +1,42 @@
+			/// <summary>
+            /// Specifies whether the server may enable or disable privileges and groups that the client's security context may include.
+            /// </summary>
+            internal byte effectiveOnly;
```
use `BOOLEAN`

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.SECURITY_QUALITY_OF_SERVICE.cs:
```diff
@@ -0,0 +1,42 @@
+			/// <summary>
+            /// Specifies the information given to the server about the client, and how the server may represent, or impersonate, the client.
+            /// </summary>
+            [MarshalAs(UnmanagedType.I4)]
```
This is redundant and might(?) make this unblittable. If you're going to use the field, define the enum. https://msdn.microsoft.com/en-us/library/windows/desktop/aa379572.aspx.  Even if you don't explicitly use it, consider defining it anyway for clarity.

keeratsingh replied:
I added this file as it was being used as part of the Framework implementation, since it is not being used as part of the CoreFx implementation, do you suggest removing this file altogether ?

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.SECURITY_QUALITY_OF_SERVICE.cs:
```diff
@@ -0,0 +1,42 @@
+            /// Specifies whether the server is to be given a snapshot of the client's security context (called static tracking), or is to be
+			/// continually updated to track changes to the client's security context (called dynamic tracking).
+            /// </summary>
+            internal byte contextDynamicTrackingMode;
```
Again, consider, defining. https://msdn.microsoft.com/en-us/library/cc234317.aspx

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            // ensure we have validated and normalized the path before
+            AssertPathFormat(path);
+
+            string formatPath = @"\??\UNC\{0}\{1}";
```
You should check the format of the given path. `PathInternal.IsDeviceUnc()` will see if you're already in the format you're trying to generate here. If so, you just need to switch the prefix to `\??`. (e.g. the path could already be `\\.\UNC\...` or `\\?\UNC\....`)

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+        // 2. GetFullPathName API of kernel32 does not accept paths with length (in chars) greater than 32766
+        //  (32766 is actually Int16.MaxValue - 1, while (-1) is for NULL termination)
+        // We must check for the lowest value between the the two
+        private const int MaxWin32PathLength = short.MaxValue - 1;
```
For the record, this isn't technically correct, but is unlikely to cause grief. `\??\UNC\` gets turned into `\Device\Mup\` under the covers. Null termination isn't a thing either once you're in UNICODE_STRING format.

Ideally you'd just let Windows choke and deal with the returned error rather than try to second guess here.

keeratsingh replied:
Getting rid of this check.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            Debug.Assert(path.Length <= MaxWin32PathLength);
+            Debug.Assert(path.IndexOfAny(s_invalidPathChars) < 0);
+            Debug.Assert(path.StartsWith(@"\\", StringComparison.OrdinalIgnoreCase));
+            Debug.Assert(!path.StartsWith(@"\\.\", StringComparison.Ordinal));
```
You shouldn't care about this. You're not checking for `\\?\`, which is the same thing.

keeratsingh replied:
Removed Assert statement.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            Debug.Assert(!path.StartsWith(@"\\.\", StringComparison.Ordinal));
+        }
+
+        // SQLBUVSTS01 bugs 192677 and 193221: we cannot use System.IO.Path.GetFullPath for two reasons:
```
None of this is true in Core.

keeratsingh replied:
Code removed.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            }
+
+            // make sure path is a UNC path
+            if (!path.StartsWith(@"\\", StringComparison.OrdinalIgnoreCase))
```
The right thing to check here is `!(PathInternal.IsDeviceUnc(path) || (path.StartsWith(@"\\") && !PathInternal.IsDevice(path)))`

JeremyKuhne replied:
You can even add a helper to PathInternal for IsUnc() with this logic if you want.

keeratsingh replied:
@JeremyKuhne 
Adding helper to PathInternal requires including file `$(CommonPath)\CoreLib\System\IO\PathInternal.Windows.cs`
This file gives a compilation error `cannot convert from 'string' to 'System.ReadOnlySpan<char>'` at the line below:
> https://github.com/dotnet/corefx/blob/master/src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs#L133

To resolve the error I can explicitly convert the `string` to `span`, any suggestions/comments ?
`if (IsPartiallyQualified(path.AsSpan()) || IsDevice(path.AsSpan()))`

keeratsingh replied:
> You can even add a helper to PathInternal for IsUnc() with this logic if you want.

Did you mean the method [IsDeviceUNC()](https://github.com/dotnet/corefx/blob/1b9cf748834740ad843828384cfd6967e991b115/src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs#L164) ?

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            }
+
+            // check for the path length before we normalize it with GetFullPathName
+            if (path.Length > MaxWin32PathLength)
```
As mentioned above, I wouldn't bother. Let GetFullPath give you an error.

keeratsingh replied:
Code removed.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+        // * it requires PathDiscovery permissions, which is unnecessary for SqlFileStream since we are dealing with network path
+        // * it is limited to 260 length while in our case file path can be much longer
+        // To overcome the above limitations we decided to use GetFullPathName function from kernel32.dll
+        [ResourceExposure(ResourceScope.Machine)]
```
I don't think these are useful- @stephentoub ?

JeremyKuhne replied:
@ do we need these?

---
In reply to: [190732837](https://github.com/dotnet/corefx/pull/29879#discussion_r190732837) [](ancestors = 190732837)

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            }
+
+            // GetFullPathName does not check for invalid characters so we still have to validate them before
+            if (path.IndexOfAny(s_invalidPathChars) >= 0)
```
I'd recommend not checking for invalid chars, we don't any more. I can live with this staying, however.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+            //-----------------------------------------------------------------
+
+            // normalize the path
+            path = SafeGetFullPathName(path);
```
As mentioned in the method, get rid of the helper and just call `Path.GetFullPath()`.

keeratsingh replied:
Changed method call to to `System.IO.Path.GetFullPath(path)`

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+
+            // CONSIDER: is this a precondition validation that can be done above? Or must the path be normalized first?
+            // after normalization, we have to ensure that the path does not attempt to refer to a root device, etc.
+            if (path.StartsWith(@"\\.\", StringComparison.Ordinal))
```
This isn't correct. `\\?\` is the same as `\\.\`, except it gets around MaxPath checks in Win32 and avoids normalization. If you're checking for `IsUnc` (as described above) _after_ normalizing you can be positive that you have a UNC, not some other device path. Move the IsUnc() check to this position and you'll have the intended behavior.

keeratsingh commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,979 @@
+        // </summary>
+        [ResourceExposure(ResourceScope.Machine)]
+        [ResourceConsumption(ResourceScope.Machine)]
+        static private string SafeGetFullPathName(string path)
```
Method removed.

keeratsingh commented:
Addressed review feedback:
- Removed CAS-Related code
- Created singe file SqlFileStream.UnSupported.cs to be included in Unix and UAP builds.
- Updated code to use C# keywords for types
- Reverted back to using sizeof instead of Marshal.SizeOf in Interop.NtCreateFile.cs
- Updated code to match casing of Microsoft Docs
- Converted tabs to spaces
- Renamed file Interop.Access.cs to Interop.IoControlCodeAccess.cs
- Updated enum Access to IoControlCodeAccess and added [Flags] attribute.
- Removed file Interop.SECURITY_QUALITY_OF_SERVICE.cs as it is not being used as part of the CoreFx implementation
- Removed helper method SafeGetFullPathName, instead updaed call to System.IO.Path.GetFullPath
- Updated call to IsPartiallyQualified method in PathInternal.Windows to use path.AsSpan() argument.

Pending:
- Update csproj file and code to not explicitly suppress BCL0015 warning in code.

keeratsingh commented:
@dotnet-bot test OSX x64 Debug Build

keeratsingh commented:
@JeremyKuhne Did you get a chance to review the EA info code ?

stephentoub commented:
@JeremyKuhne, @keeratsingh, what's the status of this PR?

JeremyKuhne commented:
@keeratsingh Sorry, missed the ping. I'll make sure to review tomorrow.

saurabh500 commented on src/System.Data.SqlClient/ref/System.Data.SqlClient.cs:
```diff
@@ -866,3 +866,51 @@ public sealed partial class OperationAbortedException : System.SystemException
+{
+    public sealed partial class SqlFileStream : System.IO.Stream
+    {
+        public SqlFileStream
```
Nit: This declaration should be in a single line, following the convention of the file.

saurabh500 commented on src/System.Data.SqlClient/ref/System.Data.SqlClient.cs:
```diff
@@ -866,3 +866,51 @@ public sealed partial class OperationAbortedException : System.SystemException
+            )
+        { }
+
+        public SqlFileStream
```
Nit: This declaration should be in a single line, following the convention of the file.

saurabh500 commented on src/System.Data.SqlClient/ref/System.Data.SqlClient.cs:
```diff
@@ -866,3 +866,51 @@ public sealed partial class OperationAbortedException : System.SystemException
+            )
+        { }
+
+        public string Name { get { throw null; } }
```
Nit: you can remove the new lines between each method.

saurabh500 commented on src/System.Data.SqlClient/src/System.Data.SqlClient.csproj:
```diff
@@ -229,6 +229,64 @@
+    <Compile Include="System\Data\SqlClient\SqlFileStream.Unsupported.cs" />
+  </ItemGroup>
+  <!-- Windows Non-UAP dependencies -->
+  <ItemGroup Condition="('$(TargetGroup)' == 'netstandard' OR '$(TargetGroup)' == 'netcoreapp') AND '$(TargetsWindows)' == 'true'">
```
Isn't this condition same as line 291? Is there any particular reason, you have duplicated this ItemGroup Condition?

saurabh500 commented on src/System.Data.SqlClient/src/System.Data.SqlClient.csproj:
```diff
@@ -492,7 +551,7 @@
     </EmbeddedResource>
   </ItemGroup>
-  <ItemGroup  Condition="'$(TargetGroup)' == 'netcoreapp'">
+  <ItemGroup Condition="'$(TargetGroup)' == 'netcoreapp'">
```
Nit: Indentation

keeratsingh replied:
Removed extra space in alignment with other tags.

saurabh500 commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,801 @@
+
+            IntPtr pbBuffer = IntPtr.Zero;
+            RuntimeHelpers.PrepareConstrainedRegions();
+            try
```
We could do away with the try{}finally{} block and have the code in the finally block as-is. 
There are no thread abort in .Net Core.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -33,7 +36,23 @@ internal partial class NtDll
+            ObjectAttributes objectAttributes = ObjectAttributes.OBJ_CASE_INSENSITIVE
+            )
+        {
+                return ( CreateFile(eaBuffer: new Microsoft.Win32.SafeHandles.SafeFileHandle(IntPtr.Zero, true), eaLength: 0, path, rootDirectory,createDisposition, desiredAccess, shareAccess, fileAttributes, createOptions, objectAttributes) );
```
Don't do this- you're adding an allocation for every directory to all file enumerations on Windows. Switch back to null and pointer- I'll give a comment on how to deal with your new overload.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -59,15 +78,19 @@ internal partial class NtDll
                     CreateOptions: createOptions,
-                    EaBuffer: null,
-                    EaLength: 0);
+                    EaBuffer: eaBuffer,
```
You can manually ref count here to keep the other paths allocation free and the generated P/Invoke clean (via DangerousAddRef, DangerousGetHandle, etc).  Here is one example of how to ref count: [https://github.com/dotnet/corefx/blob/e7723074a7ee7cfbc1da516108d2ad85082d77d4/src/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs#L65](https://github.com/dotnet/corefx/blob/e7723074a7ee7cfbc1da516108d2ad85082d77d4/src/System.Data.Odbc/src/Common/System/Data/ProviderBase/DbBuffer.cs#L65)

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.FILE_FULL_EA_INFORMATION.cs:
```diff
@@ -0,0 +1,45 @@
+        /// <a href="https://docs.microsoft.com/en-us/windows-hardware/drivers/ddi/content/wdm/ns-wdm-_file_full_ea_information">FILE_FULL_EA_INFORMATION</a> structure.
+        /// Provides extended attribute (EA) information. This structure is used primarily by network drivers.
+        /// </summary>
+        [StructLayoutAttribute(LayoutKind.Sequential)]
```
Nit: Structs are sequential by default. We will add them anyway sometimes to avoid warnings about naming if they pop up.

JeremyKuhne commented:
I've added some new comments. What I need to see to unblock this is NtCreateFile to be undone to it's prior state- I've given inline comments on how to do this. (Right now these changes will regress file enumeration allocations.)

What I'd also like to see is an issue opened for follow up (i.e. doesn't have to block this PR) on fixing the way the code is handling passing in the extended attribute information. It is way too complicated / inefficient. The data being passed can be created on the stack- context can't be over 255 bytes so it should all easily fit. Fixing the innefficiencies will also make the code much easier to follow. Tag me on the issue as a support contact and mark it up-for-grabs if you can't get to it. 

Please don't hesitate to ping me if I don't respond in a couple days of any updates you make. I sometimes lose things in the shuffle. :)

For the issue:


### Setting EA attributes not efficient

SqlFileStream::FileFullEaInformation does more work than it needs to.

- Can stackalloc a buffer to contain the data- no native (or managed) heap allocations needed
- Will avoid needing to wrap with SafeHandle, use _any_ marshalling methods.
- Can use FILE_FULL_EA_INFORMATION* to easily set offset, flags, lengths
- Should store EaName constant as a byte array of ASCII bytes rather than converting every time

keeratsingh commented:
@JeremyKuhne The requested changes have been addressed in the latest commits. If everything seems in place could you kindly approve the PR.

danmosemsft commented on src/Common/src/Interop/Windows/kernel32/Interop.CTL_CODE.cs:
```diff
@@ -0,0 +1,44 @@
+            /// </summary>
+            byte access )
+        {
+            if (function > 4095)
```
Can you comment where 4095 comes from? I don't see 0xFFF in the linked page.

keeratsingh replied:
This value comes from this [documentation](https://docs.microsoft.com/en-us/windows-hardware/drivers/develop/how-to-select-and-configure-the-device-fundamental-tests), I will add a comment in the code as well.

>MaxFunctionCode: Specifies the maximum value of the FunctionCode field in the FSCTLs (or IOCTLs for IOCTL tests). The maximum possible value is 4095.

danmosemsft commented on src/Common/src/Interop/Windows/kernel32/Interop.GetFullPathName.cs:
```diff
@@ -0,0 +1,48 @@
+                /// <summary>
+                /// A pointer to a buffer that receives the null-terminated string for the drive and path.
+                /// </summary>
+                System.Text.StringBuilder buffer,
```
Nit, `using System.Text;`

keeratsingh replied:
I have removed this initially added file `Interop.GetFullPathName.cs` as is it no longer being used.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.CTL_CODE.cs:
```diff
@@ -0,0 +1,48 @@
+        internal unsafe static uint CTL_CODE
+        (
+            /// <summary>
+            /// Identifies the device type. This value must match the value that is set in the DeviceType member of the driver's DEVICE_OBJECT structure.
```
Arguments should be described via `/// <param name="deviceType"></param>`, etc in the block with the summary.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.CTL_CODE.cs:
```diff
@@ -0,0 +1,48 @@
+            /// </summary>
+            byte access )
+        {
+            /// <summary>
```
Please use regular comments inline with code (i.e. `//`)

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.DeviceIoControl.cs:
```diff
@@ -0,0 +1,28 @@
+    {
+        [DllImport(Libraries.Kernel32, CharSet = CharSet.Unicode, SetLastError = true)]
+        [ResourceExposure(ResourceScope.Machine)]
+        internal static extern bool DeviceIoControl
```
Please remove all of the [ResourceExposure] attributes.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.IoControlCodeAccess.cs:
```diff
@@ -0,0 +1,42 @@
+        /// Indicates the type of access that a caller must request when opening the file object that represents the device (see IRP_MJ_CREATE).
+        /// </summary>
+        
+        [Flags]
```
Nit: empty line.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Method.cs:
```diff
@@ -0,0 +1,43 @@
+            /// of the input and output buffers that were specified to DeviceIoControl or IoBuildDeviceIoControlRequest, without validating or mapping them.
+            /// </summary>
+            METHOD_NEITHER
+        };
```
>; [](start = 9, length = 1)

Nit: extraneous semicolon

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.Method.cs:
```diff
@@ -0,0 +1,43 @@
+        /// <a href="https://docs.microsoft.com/en-us/windows-hardware/drivers/kernel/buffer-descriptions-for-i-o-control-codes">TransferType</a>.
+        /// Indicates how the system will pass data between the caller of DeviceIoControl (or IoBuildDeviceIoControlRequest) and the driver that handles the IRP.
+        /// </summary>
+        public enum Method
```
It would be better to name this TransferType to match what it is used for. Or perhaps IoControlTransferType.

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -33,7 +33,24 @@ internal partial class NtDll
+        }
+
+        internal unsafe static (int status, IntPtr handle) CreateFile(
+            void* eaBuffer,
```
Put new overload parameters at the end (`eaBuffer` & `eaLength`);

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,698 @@
+            string formatPath = string.Empty;
+            string uniqueId = string.Empty;
+
+            formatPath = System.IO.PathInternal.IsDeviceUNC(path.AsSpan()) ? @"\??" : @"\??\UNC\{0}\{1}";
```
nit: there is an implicit conversion to span from string, you shouldn't need .AsSpan(). Additionally you're not building it right if it is a device path already.  If it is `\\.\UNC\Server\Share\SomePath` you need to change `\\.` to `\??`, not trim off `\\`. You also still need to add your uniqueId.

keeratsingh commented:
@dotnet-bot Test NETFX x86 Release Build

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -50,24 +53,28 @@ internal partial class NtDll
-                    out IntPtr handle,
-                    desiredAccess,
-                    ref attributes,
+                    FileHandle: out IntPtr handle,
```
nit: These aren't necessary, in fact you're missing one for statusBlock. The only reason I added them later was to be explicit about what null was (AllocationSize:).  In earlier versions of C# you had to name all of the arguments after named one. This really should only have just that one explicit- it helps when reading the code when explicitly calling out what arguments are when passing things like null, true, false, and other unnamed constant values. (So you don't have to look at the called method to understand what the value is.)

keeratsingh replied:
Changed code to only one explicit `AllocationSize` as the other arguments are self descriptive.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.CTL_CODE.cs:
```diff
@@ -0,0 +1,33 @@
+            byte access )
+        {
+            
+            // <a href="https://docs.microsoft.com/en-us/windows-hardware/drivers/develop/how-to-select-and-configure-the-device-fundamental-tests">MaxFunctionCode</a> method.
```
Tags like this (e.g. `<a>` have no meaning in a regular comment. The URL will still get highlighted, however.

keeratsingh replied:
Removed tag `<a>`

JeremyKuhne commented on src/Common/src/Interop/Windows/NtDll/Interop.NtCreateFile.cs:
```diff
@@ -33,7 +33,10 @@ internal partial class NtDll
+            ObjectAttributes objectAttributes = ObjectAttributes.OBJ_CASE_INSENSITIVE,
+            void* eaBuffer = null,
+            uint eaLength = 0
+           )
```
nit: please keep the trailing paren on the line above.

JeremyKuhne commented on src/Common/src/Interop/Windows/kernel32/Interop.CTL_CODE.cs:
```diff
@@ -0,0 +1,33 @@
+            ushort deviceType,
+            ushort function,
+            byte method,
+            byte access )
```
nit: extra space

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Windows.cs:
```diff
@@ -0,0 +1,685 @@
+        };
+
+        public SqlFileStream
+            (
```
I don't see this sort of indenting for parameter parenthesis in the other files in this assembly. I know the original source files were like this, but it would be worth following the CoreFX styles and keep the parens next to the arguments. There aren't many arguments here so they can go on one line. Also, unless there is a conflict, you can remove System.IO from the type names.

keeratsingh replied:
Updated to normal styling as requested.

JeremyKuhne commented on src/System.Data.SqlClient/src/System/Data/SqlClient/SqlFileStream.Unsupported.cs:
```diff
@@ -0,0 +1,55 @@
+    public sealed partial class SqlFileStream : System.IO.Stream
+    {
+        public SqlFileStream
+            (
```
Commented elsewhere. Can you please pull the arguments into our normal styling? Also it looks like you have more usings than needed in this file.

keeratsingh replied:
Updated to normal styling as requested.

keeratsingh commented on src/Common/src/Interop/Windows/kernel32/Interop.DeviceIoControl.cs:
```diff
@@ -0,0 +1,27 @@
+using Microsoft.Win32.SafeHandles;
+using System;
+using System.Runtime.InteropServices;
+using System.Runtime.Versioning;
```
Must have been carried over, removing the `using System.Runtime.Versioning;`

keeratsingh commented on src/Common/src/CoreLib/System/IO/PathInternal.Windows.cs:
```diff
@@ -130,7 +130,7 @@ internal static string EnsureExtendedPrefix(string path)
             // shimming method. (Or making a change that doesn't impact normalization, such as adding a filename to a
             // normalized base path.)
-            if (IsPartiallyQualified(path) || IsDevice(path))
+            if (IsPartiallyQualified(path.AsSpan()) || IsDevice(path.AsSpan()))
```
Undid the changes.

keeratsingh commented:
@dotnet-bot Test Windows x64 Debug Build

keeratsingh commented:
@dotnet-bot Test Linux x64 Release Build

JeremyKuhne commented:
Thanks for the follow-up @keeratsingh!

keithnicholson commented:
Wonderful!  Thank you!  Looking forward to the future release.
